File: backend/schemas/data_health.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

HealthStatus = Literal["ok", "info", "warning", "error"]


class DataHealthItem(BaseModel):
    scope: str
    check_name: str
    status: HealthStatus
    message: str
    error_code: str | None = None
    path: str | None = None
    record_type: str | None = None
    record_id: int | None = None
    project_id: int | None = None
    suggestion: str | None = None
    is_checked_file: bool = Field(default=False, exclude=True)
    is_missing_file: bool = Field(default=False, exclude=True)


class OrphanFileItem(BaseModel):
    project_id: int
    path: str
    size_bytes: int
    suggestion: str = "该文件未被数据库记录引用。请确认不是人工保留文件后再手动处理。"


class DataHealthSummary(BaseModel):
    ok_count: int = 0
    info_count: int = 0
    warning_count: int = 0
    error_count: int = 0
    checked_file_count: int = 0
    missing_file_count: int = 0
    orphan_file_count: int = 0
    orphan_file_size_bytes: int = 0
    temp_file_count: int = 0
    temp_size_bytes: int = 0
    project_count: int = 0
    backup_count: int = 0
    export_count: int = 0
    restore_count: int = 0


class DataHealthGroupSummary(BaseModel):
    error: int = 0
    warning: int = 0
    info: int = 0


class SystemHealthResult(BaseModel):
    status: HealthStatus
    generated_at: datetime
    app_data_path: str
    summary: DataHealthSummary
    grouped_summary: dict[str, DataHealthGroupSummary] = Field(default_factory=dict)
    items: list[DataHealthItem] = Field(default_factory=list)


class ProjectHealthResult(BaseModel):
    project_id: int
    project_name: str
    status: HealthStatus
    generated_at: datetime
    project_path: str
    summary: DataHealthSummary
    grouped_summary: dict[str, DataHealthGroupSummary] = Field(default_factory=dict)
    items: list[DataHealthItem] = Field(default_factory=list)
    orphan_files: list[OrphanFileItem] = Field(default_factory=list)


class BackupRecordHealthResult(BaseModel):
    status: HealthStatus
    generated_at: datetime
    summary: DataHealthSummary
    grouped_summary: dict[str, DataHealthGroupSummary] = Field(default_factory=dict)
    items: list[DataHealthItem] = Field(default_factory=list)


class ExportRecordHealthResult(BaseModel):
    status: HealthStatus
    generated_at: datetime
    summary: DataHealthSummary
    grouped_summary: dict[str, DataHealthGroupSummary] = Field(default_factory=dict)
    items: list[DataHealthItem] = Field(default_factory=list)


class OrphanFileScanResult(BaseModel):
    project_id: int
    status: HealthStatus
    generated_at: datetime
    summary: DataHealthSummary
    grouped_summary: dict[str, DataHealthGroupSummary] = Field(default_factory=dict)
    orphan_files: list[OrphanFileItem] = Field(default_factory=list)


class TempCleanupResult(BaseModel):
    status: HealthStatus
    deleted_file_count: int
    deleted_dir_count: int
    freed_bytes: int
    errors: list[str] = Field(default_factory=list)


class DataSafetySummary(BaseModel):
    app_data_path: str
    database_exists: bool
    projects_dir_exists: bool
    backups_dir_exists: bool
    logs_dir_exists: bool
    temp_dir_exists: bool
    project_count: int
    backup_count: int
    export_count: int
    restore_count: int
    app_data_writable: bool


class MaintenanceReportResult(BaseModel):
    status: HealthStatus
    generated_at: datetime
    report_markdown: str
    system_health: SystemHealthResult

File: frontend/src/api/dataHealth.ts
import { apiGet, apiPost } from "./client";

export type DataHealthStatus = "ok" | "info" | "warning" | "error";

export type DataHealthItem = {
  scope: string;
  check_name: string;
  status: DataHealthStatus;
  message: string;
  error_code: string | null;
  path: string | null;
  record_type: string | null;
  record_id: number | null;
  project_id: number | null;
  suggestion: string | null;
};

export type OrphanFileItem = {
  project_id: number;
  path: string;
  size_bytes: number;
  suggestion: string;
};

export type DataHealthSummary = {
  ok_count: number;
  info_count: number;
  warning_count: number;
  error_count: number;
  checked_file_count: number;
  missing_file_count: number;
  orphan_file_count: number;
  orphan_file_size_bytes: number;
  temp_file_count: number;
  temp_size_bytes: number;
  project_count: number;
  backup_count: number;
  export_count: number;
  restore_count: number;
};

export type DataHealthGroupSummary = {
  error: number;
  warning: number;
  info: number;
};

export type SystemHealthResult = {
  status: DataHealthStatus;
  generated_at: string;
  app_data_path: string;
  summary: DataHealthSummary;
  grouped_summary: Record<string, DataHealthGroupSummary>;
  items: DataHealthItem[];
};

export type ProjectHealthResult = {
  project_id: number;
  project_name: string;
  status: DataHealthStatus;
  generated_at: string;
  project_path: string;
  summary: DataHealthSummary;
  grouped_summary: Record<string, DataHealthGroupSummary>;
  items: DataHealthItem[];
  orphan_files: OrphanFileItem[];
};

export type OrphanFileScanResult = {
  project_id: number;
  status: DataHealthStatus;
  generated_at: string;
  summary: DataHealthSummary;
  grouped_summary: Record<string, DataHealthGroupSummary>;
  orphan_files: OrphanFileItem[];
};

export type TempCleanupResult = {
  status: DataHealthStatus;
  deleted_file_count: number;
  deleted_dir_count: number;
  freed_bytes: number;
  errors: string[];
};

export type DataSafetySummary = {
  app_data_path: string;
  database_exists: boolean;
  projects_dir_exists: boolean;
  backups_dir_exists: boolean;
  logs_dir_exists: boolean;
  temp_dir_exists: boolean;
  project_count: number;
  backup_count: number;
  export_count: number;
  restore_count: number;
  app_data_writable: boolean;
};

export type MaintenanceReportResult = {
  status: DataHealthStatus;
  generated_at: string;
  report_markdown: string;
  system_health: SystemHealthResult;
};

export function runSystemHealthCheck(): Promise<SystemHealthResult> {
  return apiGet<SystemHealthResult>("/api/system/health-check", { timeoutMs: 120_000 });
}

export function runProjectHealthCheck(projectId: number): Promise<ProjectHealthResult> {
  return apiGet<ProjectHealthResult>(`/api/projects/${projectId}/health-check`, { timeoutMs: 120_000 });
}

export function scanProjectOrphanFiles(projectId: number): Promise<OrphanFileScanResult> {
  return apiGet<OrphanFileScanResult>(`/api/projects/${projectId}/orphan-files`, { timeoutMs: 120_000 });
}

export function cleanupTempFiles(): Promise<TempCleanupResult> {
  return apiPost<TempCleanupResult>("/api/system/cleanup-temp", undefined, { timeoutMs: 120_000 });
}

export function getDataSafetySummary(): Promise<DataSafetySummary> {
  return apiGet<DataSafetySummary>("/api/system/data-safety-summary");
}

export function buildMaintenanceReport(): Promise<MaintenanceReportResult> {
  return apiGet<MaintenanceReportResult>("/api/system/maintenance-report", { timeoutMs: 120_000 });
}
